packages/element/src/create-interpolate-element.js

```diff
@@ -0,0 +1,239 @@
+		}
+	);
+};
+
+/**
+ * This reorders the conversionMap so that it's entries match the order of
+ * elements in the string.
+ *
+ * This is necessary because the parser is order sensitive due to the potential
+ * for nested elements (eg. <a>Some linked <em>and emphasized</em></a> string).
+ * This ensures that the parsing will be done correctly yet still allow for the
+ * consumer not to worry about order in the map.
+ *
+ * @param interpolatedString {string}  The string being parsed.
+ * @param conversionMap      {Array}   The map being reordered
+ *
+ * @return {Array}  The new map in the correct order for the tags in the string.
+ */
+const reorderMapByElementOrder = ( interpolatedString, conversionMap ) => {
+	// if length of map is only one then we can just return as is.
+	if ( conversionMap.length === 1 ) {
+		return conversionMap;
+	}
+	return conversionMap.sort( ( [ tagA ], [ tagB ] ) => {
+		tagA = `<${ tagA }`;
+		tagB = `<${ tagB }`;
+		return interpolatedString.indexOf( tagA ) > interpolatedString.indexOf( tagB );
+	} );
+};
+
+/**
+ * This function creates an interpolated element from a passed in string with
+ * specific tags matching how the string should be converted to an element via
+ * the conversion map value.
+ *
+ * @example
+ * For example, for the given string:
+ *
+ * "This is a <span>string</span> with <a>a link</a>, a self-closing
+ * <CustomComponentB/> tag and a plain value <custom value/>"
+ *
+ * You would have something like this as the conversionMap value:
+ *
+ * ```js
+ * {
+ *     span: { tag: CustomComponent, props: {} },
+ *     a: { tag: 'a', props: { href: 'https://github.com' } },
+ *     CustomComponentB: { tag: CustomComponentB, props: {} },
+ *     'custom value': { value: 'custom value' },
+ * }
+ * ```
+ *
+ * @param {string}  interpolatedString  The interpolation string to be parsed.
+ * @param {Object}  conversionMap       The map used to convert the string to
+ *                                      a react element.
+ *
+ * @return {Element}  A react element.
+ */
+const createInterpolateElement = ( interpolatedString, conversionMap ) => {
+	keyIndex = -1;
+
+	if ( ! isValidConversionMap( conversionMap ) ) {
+		return interpolatedString;
+	}
+
+	// verify that the object isn't empty.
+	conversionMap = Object.entries( conversionMap );
+	if ( conversionMap.length === 0 ) {
+		return interpolatedString;
+	}
+
+	return createElement(
+		Fragment,
+		{},
+		recursiveCreateElement(
+			interpolatedString,
+			reorderMapByElementOrder( interpolatedString, conversionMap )
+		),
+	);
+};
+
+/**
+ * Validate conversion map.
+ *
+ * A map is considered valid if it's an object and does not have length.
+ *
+ * @param conversionMap {Object}  The map being validated.
+ *
+ * @return boolean  True means the map is valid.
+ */
+const isValidConversionMap = ( conversionMap ) => {
+	return typeof conversionMap === 'object' &&
+		typeof conversionMap.length === 'undefined';
```

Comment:
ya I was trying to avoid using lodash in here.  Otherwise I'd just use `lodash.isPlainObject`.

Reply:
this isn't actually detecting if we have a valid conversion map; it's determining if we're an object without a length property.

```js
isValidConversionMap( new Map() ) // true
isValidConversionMap( {} ) // true
isValidConversionMap( Object.create(null) ) // true
isValidConversionMap( JSON ) // true
```

If we're going to validate we probably want to make more certain assurances and even give a more specific name to what we're doing. If I read `isValid` I don't really know how to know if the function is correct or finished.

```js
const conversionMapError = s => `createInterpolateComponent: second argument should be { [string]: Element } but ${ s }`; 

const matchesObjectOfElements = o => {
	if ( 'object' !== typeof o ) {
		return [ false, conversionMapError( `found ${ typeof o }` ) ];
	}

	o.getOwnPropertyNames().forEach( name => {
		if ( ! ( o[name] instanceof Element ) ) {
			return [ false, conversionMapError( `found key ${ name } of type ${ typeof o[name] }` ) ];
		}
	} );

	return [ true ];
}

const validateConversionMap = map => {
	const [ isObjectOfElements, error ] = matchesObjectOfElements( map );
	if ( IS_DEBUG && ! isObjectOfElements ) {
		throw new TypeError( error );
	}
}

const isValidConversionMap = map => {
	try {
		validateConversionMap( map );
		return true;
	} catch ( e ) {
		return false;
	}
}
```

This is verbose and ugly but it more closely follows the constraints we are imposing on the system - it's verifying that the things given won't break in the code we write.

By splitting the part that throws from the part that checks we can provide different levels of help when things are invalid. It's probably fine to skip returning `true` and `false` entirely and rely on throwing the errors. If there's any decent fallback behavior to use then we can try and to that with the boolean catch on `false` but otherwise I'm a fan of letting it crash as early as possible to make it harder to miss the bug.

Reply:
Truly valid conversion maps can only be determined when trying to insert the elements. Another approach is to throw when adding the component into our stack or output. That's less clean but more complete. That approach is also complementary with this approach: validate structure up-front and validate data during the loop.